### ISSUE TITLE
Backlinks support

### ISSUE THREAD

**Issue by heitorPB** [getzola/zola#1496]
Is there support for backlinks? That is, a list of pages that link to the current one. There is a very awesome example of this feature in [Andy's notes](https://notes.andymatuschak.org/About_these_notes?stackedNotes=zVFGpprS64TzmKGNzGxq9FiCDnAnCPwRU5T)

I saw a [question in the forums about this](https://zola.discourse.group/t/does-zola-support-backlinks/655), saying it is possible to add this feature, but I didn't see an update there or a mention in the docs.

The PR https://github.com/getzola/zola/pull/1424 mentions this feature as well. I took the initiative to open this issue to track this feature. I don't know rust, but I'd like to help in any way I can. Maybe with docs?

**Comment by Keats** [comment]
So this is not fully implemented. I'd like to get it into the next release because it's something I want to use myself but I can't promise.
Implementation wise, there isn't that much left to do actually but it's 99% Rust to do, the documentation will be fairly short on that. I believe we just need to iterate through the internal links of each page and populate a new `backlinks` field with either: 
1. the permalinks/title/description
2. just the path and let users pick what they want from the template via get_page/get_section

I prefer 1 for now, maybe 2 will be better in a world where Tera is more efficient and doesn't clone as much.

**Comment by heitorPB** [comment]
Let me know if there's something I can to do help here :)

**Comment by coastalwhite** [comment]
I think any implementation of this issue should be done after #1564, since it will probably end up doing a lot changes in that area of the codebase as well.

Maybe it is even an idea to do those PR’s simultaneously in #1564 and add this issue to that PR’s scope? I feel like an objection would be that that PR is starting to get quite big already (probably reaching +5000 before it is finished). Since that PR is probably going need to be part of the next major release anyway, it shouldn’t be too big of a problem.

What is your take on this @Keats?

**Comment by Keats** [comment]
Why do you think it should be done after the rewrite? Afaik it shouldn't really interact with it?

**Comment by coastalwhite** [comment]
I think PR will probably, in the process, touch on the parsing code for internal links and the `RenderContext` struct. So I suspect there may be some issues which may occur during merging and maybe even (probably and hopefully not) in the logical part if these are developed concurrently.

**Comment by boehs** [comment]
I might have a go at this provided no one else is planning to (fair warning: I have a lot of shit right now so it will take a bit but I need this functionality and know rust)

**Comment by mjf** [comment]
Do you think that once having the backlinks feature included it may be possible to achieve creating of something functionaly similar to the [Andy's notes](https://notes.andymatuschak.org/About_these_notes?stackedNotes=zVFGpprS64TzmKGNzGxq9FiCDnAnCPwRU5T) with Zola (and perhaps with some bunch of additional JS)? I ask because I like the concept very much personally...

Anyway, if not, an additional section with backlinks below a content will be extremely helpful for me, so I am looking forward to the backlinks feature very much! Thanks for your great work! :heart:

**Comment by mjf** [comment]
Have you considered grouping of the backlinks based on for example categories?

Motivation: An overview could be added with counts of content backlinking from each category followed by for example with an ordered list of categories with sublists of backlinks for each...

... Just an petite idea to think about while the backlinking feature is not yet implemented.

**Comment by Keats** [comment]
> Do you think that once having the backlinks feature included it may be possible to achieve creating of something functionaly similar to the Andy's notes with Zola (and perhaps with some bunch of additional JS)? I ask because I like the concept very much personally...

That should be easy yes, and it's the point of backlinks in the first place.

> Have you considered grouping of the backlinks based on for example categories?

Probably not built-in, but it could be done by the user I believe.

**Comment by mjf** [comment]
>> Have you considered grouping of the backlinks based on for example categories?

> Probably not built-in, but it could be done by the user I believe.

If you could then provide a simple snippet in the docs to start with, it would be greatly appreciated (by me). Thank you.

**Comment by Keats** [comment]
I'm not sure it makes sense to be in the docs but I'll make sure to try it when the feature is added

**Comment by heitorPB** [comment]
> Do you think that once having the backlinks feature included it may be possible to achieve creating of something functionaly similar to the [Andy's notes](https://notes.andymatuschak.org/About_these_notes?stackedNotes=zVFGpprS64TzmKGNzGxq9FiCDnAnCPwRU5T) with Zola (and perhaps with some bunch of additional JS)? I ask because I like the concept very much personally...

Yes. There is one for Hugo, so it should be straightforward to make a template for Zola: https://github.com/jethrokuan/cortex

**Comment by mjf** [comment]
Great! I definitely want to use the backlinks for both my new web site's posts and for my notes (I even have already started to organize them - what a mess it is!). I think I will make use of them in other areas too (not sure yet). Anyway, I hope backlinks feature arrives soon... :smile:

**Comment by Keats** [comment]
The first part (collecting internal links) is already done, they just need to be exposed in a way that makes sense. The next release will have a focus on refactoring/cleaning up though so I don't know whether I'll add it for this release or the one after.

**Comment by aadimator** [comment]
@Keats Can you kindly share a demo of how to use/collect backlinks? That'd be great.

**Comment by Keats** [comment]
It's not implemented yet. The collection of backlinks is/will be automatic.

**Comment by Keats** [comment]
Backlinks have been added to the `next` branch.
It's not documented yet as it might change but basically every page/section gets a backlinks attribute that is a vec of `{ permalink: &str, title: &'a Option<String>}`.

If people can try and report if they are missing things/issues that would be great.

**Comment by Keats** [comment]
Any feedback on it so far?

**Comment by Keats** [comment]
For my own tiny personal knowledge base, it does everything I need FWIW

**Comment by timharek** [comment]
Awesome! I just tested it and it works fine for templates, but it seems that I cannot access it with a custom shortcode.

Is that expected behavior? 💭 @Keats

**Comment by Keats** [comment]
Yes it's normal. Links collection and shortcode rendering happen at the same time so backlinks can't be available in shortcodes.

**Comment by timharek** [comment]
Should this issue be labeled with `done in pr` now? 💭

**Comment by pinpox** [comment]
Is this merged in the v0.16.0? If so, is it documented yet?

**Comment by Keats** [comment]
Looks like I forgot to add the documentation >_>

**Comment by pinpox** [comment]
> Looks like I forgot to add the documentation >_>

I'm subscribed to this issue, please leave a comment when it get's added so I get a notification.

**Comment by Keats** [comment]
I've just pushed it, it will be live any time now: https://github.com/getzola/zola/commit/15a523b62474d576273b1a1397a3dc6be017321f

**Comment by thezeroalpha** [comment]
Thanks for adding this! I'm trying to get backlinks to work on my website with version 0.16.1 (macOS), but for some reason the backlinks array seems to be empty. I'm [attaching a minimal website example](https://github.com/getzola/zola/files/9337933/backlinks-minimal-example.tgz). With this setup, I would expect e.g. "page two" to have a backlink from "page one", but there seem to be no backlinks, and the Tera context shows `"backlinks": []`. Am I doing something wrong here?

**Comment by Keats** [comment]
You need to use Zola internal links, the ones starting with @/
